src/xdp-utils.c

```diff
@@ -144,11 +149,79 @@ xdp_app_info_new (XdpAppInfoKind kind)
   return app_info;
 }
 
+char *
+_xdp_parse_app_id_from_unit_name (const char *unit)
+{
+  g_autoptr(GRegex) regex = NULL;
+  g_autoptr(GMatchInfo) match = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree char *app_id = NULL;
+
+  g_assert (g_str_has_prefix (unit, "app-"));
+
+  /*
+   * From https://systemd.io/DESKTOP_ENVIRONMENTS/ the format is one of:
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.scope
+   * app[-<launcher>]-<ApplicationID>-autostart.service -> no longer true since systemd v248
+   * app[-<launcher>]-<ApplicationID>[@<RANDOM>].service
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.slice
+   */
+  regex = g_regex_new ("^app-(?:[[:alnum:]]+\\-)?(.+?\\..+?\\..+?)"
+                       "(?:[@\\-][0-9]*)?(?:-autostart)?"
+                       "(?:\\.scope|\\.service|\\.slice)$",
+                       0, 0, &error);
+  g_assert (error == NULL);
+  if (g_regex_match (regex, unit, 0, &match))
+    {
+      const char *escaped_app_id;
+      /* Unescape the unit name which may have \x hex codes in it, e.g.
+       * "app-gnome-org.gnome.Evolution\x2dalarm\x2dnotify-2437.scope"
+       */
+      escaped_app_id = g_match_info_fetch (match, 1);
+      if (cunescape (escaped_app_id, UNESCAPE_RELAX, &app_id) < 0)
+        app_id = g_strdup ("");
+    }
+  else
+    {
+      app_id = g_strdup ("");
+    }
+
+  return g_steal_pointer (&app_id);
+}
+
+void
+get_app_id_from_pid (pid_t pid, XdpAppInfo *app_info)
```

Comment:
This call seems to unconditionally assign `app_info->id`, so it might be good to add a guard like:
```
g_return_if_fail (app_info->id == NULL);
```

Also, a bit late, but maybe `set_appid_from_pid()` is a better name, and pass `app_info` as the first argument?

Reply:
Sure, makes sense on both counts